code/modules/station_goals/dna_vault.dm

```diff
@@ -0,0 +1,190 @@
+		if(!H.myseed)
+			return
+		if(!H.harvest)// So it's bit harder.
+			user << "<span clas='warning'>Plants needs to be ready to harvest to perform full data scan.</span>" //Because space dna is actually magic
+			return
+		if(plants[H.myseed.type])
+			user << "<span class='notice'>Plant data already present in local storage.<span>"
+			return
+		plants[H.myseed.type] = 1
+		user << "<span class='notice'>Plant data added to local storage.<span>"
+
+	//animals
+	if(isanimal(target) || is_type_in_typecache(target.type,non_simple_animals))
+		if(isanimal(target))
+			var/mob/living/simple_animal/A = target
+			if(!A.healable)//simple approximation of being animal not a robot or similar
+				user << "<span class='warning'>No compatibile DNA detected</span>"
+				return
+		if(animals[target.type])
+			user << "<span class='notice'>Animal data already present in local storage.<span>"
+			return
+		animals[target.type] = 1
+		user << "<span class='notice'>Animal data added to local storage.<span>"
+
+	//humans
+	if(ishuman(target))
+		var/mob/living/carbon/human/H = target
+		if(dna[H.dna.uni_identity])
+			user << "<span class='notice'>Humanoid data already present in local storage.<span>" 
+			return
+		dna[H.dna.uni_identity] = 1
+		user << "<span class='notice'>Humanoid data added to local storage.<span>"
+
+
+/obj/item/weapon/circuitboard/machine/dna_vault
+	name = "circuit board (DNA Vault)"
+	build_path = /obj/machinery/dna_vault
+	origin_tech = "engineering=2;combat=2;bluespace=2" //No freebies!
+	req_components = list(
+							/obj/item/weapon/stock_parts/capacitor/quadratic = 5,
+							/obj/item/stack/cable_coil = 2)
+
+/obj/machinery/dna_vault
+	name = "DNA Vault"
+	desc = "Break glass in case of apocalypse."
+	icon = 'icons/obj/machines/dna_vault.dmi'
+	icon_state = "vault"
+	density = 1
+	anchored = 1
+	idle_power_usage = 5000
+	pixel_x = -32
+	pixel_y = -64
+	luminosity = 1
+	var/list/animals = list()
+	var/list/plants = list()
+	var/list/dna = list()
+
+	var/list/obj/structure/fillers = list()
+
+/obj/machinery/dna_vault/New()
+	//TODO: Replace this,bsa and gravgen with some big machinery datum
+	var/list/occupied = list()
+	for(var/direct in list(EAST,WEST,SOUTHEAST,SOUTHWEST))
+		occupied += get_step(src,direct)
+	occupied += locate(x+1,y-2,z)
+	occupied += locate(x-1,y-2,z)
+
+	for(var/T in occupied)
+		var/obj/structure/filler/F = new(T)
+		F.parent = src
+		fillers += F
+
+/obj/machinery/dna_vault/Destroy()
+	for(var/V in fillers)
+		var/obj/structure/filler/filler = V
+		filler.parent = null
+		qdel(filler)
+	. = ..()
+
+
+/obj/machinery/dna_vault/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/ui_state/state = physical_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "dna_vault", name, 250, 200, master_ui, state)
+		ui.open()
+
+/obj/machinery/dna_vault/ui_data() //TODO Make it % bars maybe
```

Comment:
yes, that's as easy as providing the 3 max values to the ui_data() as well and using a ui-bar object.